### PR TITLE
Feature/right click contextual open for ic popover menu

### PR DESCRIPTION
**Pull request by c95560**
## Summary of the changes
Adds an optional right-click contextual menu open to the ic-popover-menu. This potentially reduces the number of mouse actions a user needs to take, and could provide useful mouse x-y position information for events (e.g. right-click - open menu - add new component at mouse x-y position)

## Related issue
None known

## Checklist
- [?] New web component and react stories added to demonstrate feature. Not clear if unit tests needed. Visual style not changed.
- [y] I have manually accessibility tested any changes, if relevant. Verified existing behaviour not broken.
- [y] I have ensured any changes match the Figma component library.

### CODE REVIEW

**Comment by jd239** [comment]
Hi @c95560. Thank you for your contribution. Can you please refer to the [Contributing ](https://github.com/mi6/ic-ui-kit/blob/main/CONTRIBUTING.md) file and make updates to your files, rebase and commits were relevant.

**Review by GCHQ-Developer-530 — CHANGES_REQUESTED** [review]
A couple of console logs left in there that can now be removed. Also some styling issues that should be easily fixed by running npm run prettier:fix 😊

File: packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx [hunk 1]
```diff
@@ -37,6 +37,44 @@ export class PopoverMenu {
...
+   * if mouseAnchor is set to true. If null, then use the root element.
+   */
+  @Prop() clickEventEle: string = document.documentElement.id;
+  private clickEventEleRect: DOMRect = undefined;
+
+  /**
+   *  State for popover render position
+   */
+  @Prop() renderX: number = null;
+  @Prop() renderY: number = null;
```
Comment (GCHQ-Developer-530) [inline packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:50]: This prop will need it's own description.

File: packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx [hunk 2]
```diff
@@ -37,6 +37,44 @@ export class PopoverMenu {
   @Prop() anchor: string;
 
   /**
+   * The ID of the element that can open the popover menu
+   * if mouseAnchor is set to true. If null, then use the root element.
+   */
+  @Prop() clickEventEle: string = document.documentElement.id;
+  private clickEventEleRect: DOMRect = undefined;
```
Comment (GCHQ-Developer-530) [inline packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:44]: Please can you follow our [style guide](https://github.com/mi6/ic-ui-kit/blob/main/CONTRIBUTING.md#style-guide) and move the new private variables and props into the correct place in the file and in alphabetical order?

File: packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx [hunk 3]
```diff
@@ -37,6 +37,44 @@ export class PopoverMenu {
...
+
+  /**
+   *  State for popover render position
```
Comment (GCHQ-Developer-530) [inline packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:47]: Think this prop description should be specific to renderX

**Review by GCHQ-Developer-530 — CHANGES_REQUESTED** [review]
Couple more comments

**Review by GCHQ-Developer-530 — CHANGES_REQUESTED** [review]
Please could you also squash your commits into one's that follow [conventional commits](https://github.com/mi6/ic-ui-kit/blob/main/CONTRIBUTING.md#git-commit). You should have one for your update to react stories and another for the web-components work you've done.

**Comment by c95560** [comment]
@GCHQ-Developer-530

Noticed some issues resolving merge conflict - duplicated functions etc. 
Rebuilt branch enacting requested changes and retested.
- Fixed a console error bug related to bad anchor element DOM query.
- Added missing prop descriptions, e.g. for renderX
- Improved style by alphabetically ordering declarations of class props, functions etc
- recommitted code as a single commit via conventional commits plugin
- removed console debug messages that had been left in ic-popover-menu.tsx
- removed commented out debug code
- fixed style issues with npm run prittier:fix
- fixed (minor) merge conflicts with latest changes from main

File: packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx [hunk 4]
```diff
@@ -137,17 +133,67 @@ export class PopoverMenu {
...
+      setTimeout(this.setButtonFocus, 50);
+    }
+  }
+
   @Listen("click", { target: "document" })
-  handleClick(ev: Event): void {
-    if (this.open && this.isNotPopoverMenuEl(ev)) {
-      // If menu is open and the next click on the document is not a popover El, close the popover
+  handleClick(ev: MouseEvent): void {
+    //document.documentElement.style.backgroundColor = 'red';
```
Comment (GCHQ-Developer-530) [inline packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:160]:
```suggestion
```

File: packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx [hunk 5]
```diff
@@ -20,100 +20,96 @@ import { createPopper } from "@popperjs/core";
...
-  watchOpenHandler(): void {
-    if (this.open) {
-      if (
-        this.parentPopover !== undefined &&
-        !this.popoverMenuEls.some((menuItem) => menuItem.id)
-      ) {
-        this.popoverMenuEls.unshift(this.backButton);
-      }
+  /**
+   * If true, the ic-popover-menu will open when a MouseEvent button press occurs.
```
Comment (GCHQ-Developer-530) [inline packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:81]:
```suggestion
   * If `true`, the ic-popover-menu will open when a MouseEvent button press occurs.
```

File: packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx [hunk 6]
```diff
@@ -20,100 +20,96 @@ import { createPopper } from "@popperjs/core";
...
   /**
-   * The unique identifier for a popover submenu.
+   * The MouseEvent.button values that can trigger an ic-popover-menu open.
+   * Defaults to [2], i.e. mouse right click only
+   *
+   * 0 = left click
+   * 1 = middle click
+   * 2 = right click
+   * 3 = mouse browser back
+   * 4 = mouse browser forward
```
Comment (GCHQ-Developer-530) [inline packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:59]: Might be worth exporting these as an enum and setting the type to that as we only want to accept these numbers, not any number, and they're for the equivalent close prop

File: packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx [hunk 7]
```diff
@@ -20,100 +20,96 @@ import { createPopper } from "@popperjs/core";
...
+  @Prop() clickEventEle: string = document.documentElement.id;
 
   /**
-   * @internal The parent popover menu of a child popover menu.
+   *  absolute screen X co-ordinate for popover menu render position.
    */
-  @Prop() parentPopover?: HTMLIcPopoverMenuElement;
+  @Prop() renderX: number = null;
+  /**
+   *  absolute screen Y co-ordinate for popover menu render position.
```
Comment (GCHQ-Developer-530) [inline packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:47]:
```suggestion
   *  Absolute screen Y co-ordinate for popover menu render position.
```

File: packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx [hunk 8]
```diff
@@ -20,100 +20,96 @@ import { createPopper } from "@popperjs/core";
...
-   * @internal The parent popover menu of a child popover menu.
+   * The ID of the element that can open the popover menu
+   * if mouseAnchor is set to true. If null, then use the root element.
    */
-  @Prop() parentLabel?: string;
+  @Prop() clickEventEle: string = document.documentElement.id;
 
   /**
-   * @internal The parent popover menu of a child popover menu.
+   *  absolute screen X co-ordinate for popover menu render position.
```
Comment (GCHQ-Developer-530) [inline packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:43]:
```suggestion
   *  Absolute screen X co-ordinate for popover menu render position.
```

**Review by GCHQ-Developer-530 — CHANGES_REQUESTED** [review]

**Comment by GCHQ-Developer-530** [comment]
Thanks for updating to our conventional commit style, there's just a couple of updates to be made with the commits:
- Remove the merge commit
- Separate work out into one commit containing only the react work -  this should contain the react scope instead of ic-popover-menu something like feat(react) and then detail the work you have done
- The rest of the work can go under the web-components scope - feat(web-components) and then detail the work you've done

**Comment by c95560** [comment]
@GCHQ-Developer-530 

All prior outstanding issues should be resolved.

- Resolved additional outstanding issues with style / commented code / linting issues
- Added enum to denote allowed mouse click actions for ic-popover-menu options
- Removed merge commit
- Synced with develop branch instead of main
- Separated commits to web component and react file changes

File: packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx [hunk 9]
```diff
@@ -12,6 +12,14 @@ import {
...
 import { createPopper } from "@popperjs/core";
 
+enum IcPopoverMenuClickEnum {
+  Left = 0,
+  Middle = 1,
+  Right = 2,
+  Back = 3,
+  Forward = 4
+}
+
```
Comment (GCHQ-Developer-530) [inline packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:22]: I think this should be exported from an ic-popover-menu.types.ts file like we've done for other components (see ic-button for an example)

**Review by GCHQ-Developer-530 — CHANGES_REQUESTED** [review]

**Comment by GCHQ-Developer-530** [comment]
Sorry, it seems to be failing on prettier. Please could you run prettier:fix?

**Comment by c95560** [comment]
newest changes should rectify linting issues @GCHQ-Developer-530

**Review by GCHQ-Developer-530 — CHANGES_REQUESTED** [review]
This change seems to have caused a unit test to fail, and isn't reaching the 80% test coverage threshold. Would you be able to add some tests for the code you've written? 😊
Also noticed there's a lot of console warnings about render X and render Y about them being immutable.

File: packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx [hunk 10]
```diff
@@ -20,100 +21,87 @@ import { createPopper } from "@popperjs/core";
...
+    IcPopoverMenuClickEnum.Right,
+  ];
 
   /**
-   * @internal The level of menu being displayed.
+   * The MouseEvent.button values that can trigger an ic-popover-menu close.
+   * Defaults to `IcPopoverMenuClickEnum.Left` only.
    */
-  @Prop() submenuLevel: number = 1;
+  @Prop() mouseEventCloseButtonPressVals: IcPopoverMenuClickEnum[] = [
```
Comment (jd239) [inline packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:64]: Can this be changed to `mouseButtonClosePopover` so it's more readable

**Review by jd239 — CHANGES_REQUESTED** [review]
Apologies for the late response. First and foremost, thanks for contributing this complex functionality to the popover menu. 🎉 
A few things to amend in addition to the comments added:
- It's worth checking out the [Contribution file on git commits](https://github.com/mi6/ic-ui-kit/blob/main/CONTRIBUTING.md#git-commit) on how to format the commit messages so they pass the linting step during build.
- Rebase off the main repo to fix conflicts. 

Thanks 😃

**Comment by c95560** [comment]
I'm very short on time right now so probably won't have time to tidy this up to required standard if significant work is required.

Will revisit eventually - but more than happy to hand this off to another dev if they want to get it merged before any significant conflicts creep in.

**Comment by GCHQ-Developer-465** [comment]
> I'm very short on time right now so probably won't have time to tidy this up to required standard if significant work is required.
> 
> Will revisit eventually - but more than happy to hand this off to another dev if they want to get it merged before any significant conflicts creep in.

Hey there! Thank you for your contribution :) 

We are happy to carry on with this and update the PR. 

Thank you again, your contribution is much appreciated. We look forward to seeing more contributions!

**Comment by jd239** [comment]
This work is continues in https://github.com/mi6/ic-ui-kit/tree/feature/popover-right-click-functionality. This PR will be closed and a new PR will be created with the peer review comments.

**Pull request closed without merging**